### PR TITLE
QPIDJMS-548: support taking principle and credentials from context

### PR DESCRIPTION
**Pull request by michaelandrepearce**
Add mapping from jndi principle and credentials to username and password for the default connection factory when using provider url 
Add test case

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
@gemmellr :dog: please.

**Review by gemmellr — COMMENTED** [review]
My initial reaction doesnt have me inclined toward doing this.

Each provider has at least a couple of long standing vendor neutral ways of being passed connection credentials. E.g you are already passing the URI, and that typically lets you set them, and you can pass them when creating the connection. Existing code has to be doing something along these lines.

The previous change in QPIDJMS-542 was to permit migrations between Qpid JMS and either ActiveMQ 5 or Artemis of existing third party programs that you found relied on those providers support of naming-provider-URL based config (something I dont really like and so had previously deliberately not done; none of the other Qpid clients ever supported it). Neither of them seem to support this configuration behaviour however, which i similarly am not fond of.

I expect you can also already achieve the effect of what you want without making any changes to the Qpid JMS clients context factory, by just using the existing variable interpolation that it already supports [1] (the process is actually visible in the diff context here). E.g configuring the username+password with the appropriate variables in the url property you provide. You could also instead add the same defaulting environment property this change leveraged into the environment directly as intended, if you didnt want to reference the variables in the connection URI for some reason. Given that there are multiple ways you can achieve the effect wanted I'm then further disinclined to add this change.

[1] From https://qpid.apache.org/releases/qpid-jms-1.1.0/docs/index.html#configuring-a-jndi-initialcontext

> The property values which which constitute the connection URI, queue name, or topic name can also utilise simple ${variable} or ${variable:-default} expansion, with these resolved in order from system properties, environment variables, or the properties file / environment Hashtable.

**Comment by michaelandrepearce** [comment]
@gemmellr the issue, we are getting is that whilst ActiveMQ (Openwire) and ActiveMQ Artemis (CORE) clients seem to be honouring the credentials when set via JMS as such its vendor agnostic and secrets can be present to app securely and separately 
connectionFactory.createConnection(user,password)


where as with Qpid we are getting issue whereby connecting to secured artemis clusters we seem to be having to set the credentials on the CF itself for it to be happy with artemis when secured.  And thus looking for options how to securly and in an agnostic way (by using JNDI properties that are explicitly meant for credentials) 

putting credentials in url, isnt the best as you are essentially mixing a config item with a security item, e.g. i want the URL to be visible to ops/dev teams theres no security concerns there, but credentials clearly more sensitive. Thus why to look to use the JNDI properties that are meant for credentials.

File: qpid-jms-client/src/main/java/org/apache/qpid/jms/jndi/JmsInitialContextFactory.java [hunk 1]
```diff
@@ -80,6 +80,14 @@ public Context getInitialContext(Hashtable<?, ?> environment) throws NamingExcep
                 String expanded = expand(location, environmentCopy);
                 if (ProviderFactory.findProviderFactory(new URI(expanded)) != null) {
                     environmentCopy.put(CONNECTION_FACTORY_DEFAULT_KEY_PREFIX + REMOTE_URI_PROP, expanded);
+                    String principle = environmentCopy.containsKey(Context.SECURITY_PRINCIPAL) ? (String) environment.get(Context.SECURITY_PRINCIPAL) : System.getProperty(Context.SECURITY_PRINCIPAL);
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply qpid-jms-client/src/main/java/org/apache/qpid/jms/jndi/JmsInitialContextFactory.java:83]: fair enough can strip out the system and leave it just from jndi env.

**Comment by michaelandrepearce** [comment]
Just for your benefit, we are actually trying to switch some Kafka Connect JMS Sinks to use AMQP to get everything over to AMQP instead CORE/OPENWIRE , we are using this vendors connector.

https://github.com/lensesio/stream-reactor/tree/master/kafka-connect-jms/src/main/scala/com/datamountaineer/streamreactor/connect/jms 

Which is where the issue lays around the credentials. It seems to be not happy when we switch to QPID-JMS client and is relying on creating the connection and passing in the credentials using connectionFactory.createConnection(user,password)  but as noted when using qpid its barthing out access issues to artemis which we dont get with openwire or artemis client.

**Comment by gemmellr** [comment]
Well again, there are ways that dont require putting them in the URL, including the very one you used to implement the change under the covers.

If you are saying that the factory.createConnection(user, pass) method doesnt work and this does, then that would likely be a bug that should be investigated and fixed, but it is not a good reason to add this hack, which I still dont intend to.

For what its worth, the createConnection(user, pass) method is tested, and it operates by passing the credentials to the client innards in the same field as setting them on the factory does, since both are transferred to an intermediate object from where they actually get used, its actually the more explicitly set of the two, so I dont immediately see how this could make any difference at all to what the client does.

**Comment by michaelandrepearce** [comment]
> If you are saying that the factory.createConnection(user, pass) method doesnt work and this does, then that would likely be a bug that should be investigated and fixed, but it is not a good reason to add this hack, which I still dont intend to.
> 
Ok let me see if i can work out why this seems to not be working in the other project i referenced and see if we can make some kind of reproducer, agreed, i would also prefer to have this working as it should, just i struggled the first time to see why it wasn't working where as setting on the CF itself was. If you have any tips to where it might be causing issue for us to hunt down please share.

**Comment by gemmellr** [comment]
I dont, as I say I cant really see a reason this would alter the client behaviour at all really, during connection the different routes are essentially the same and the details are passed into the client innards in the same way at the same point in both cases, one just calls a getter to retrieve the existing value from the factory first while the other is provided it, from then they use a common impl.

**Pull request closed without merging**